autogen/agentchat/contrib/graphgroupchat.py

```diff
@@ -0,0 +1,163 @@
+import logging
+
+try:
+    import networkx as nx
+    import matplotlib.pyplot as plt
+except ImportError as e:
+    logging.fatal("Failed to import networkx or matplotlib. Try running 'pip install autogen[graphs]'")
+    raise e
+
+import autogen
+from autogen.agentchat.assistant_agent import AssistantAgent
+from autogen.agentchat.groupchat import GroupChat, Agent, ConversableAgent
+
+import random
+from typing import List, Dict
+
+
+class GraphGroupChat(GroupChat):
+    """The GraphGroupChat class inherits from the GroupChat class, and is designed to manage a group chat where the transition paths between agents are constrained by a directed graph.
+    This enables tighter control over the possible transitions between agents. Since graphs are expressive, they allow for precise control over the transitions between agents, facilitating the implementation of various organizational structures such as hierarchical, team-based, or quality review frameworks.
+    - agents: a list of participating agents. Inherited from GroupChat.
+    - messages: a list of messages in the group chat. Inherited from GroupChat.
+    - graph: a networkx graph depicting who are the next speakers available.
+    - max_round: the maximum number of rounds. Inherited from GroupChat.
+    - admin_name: the name of the admin agent if there is one. Default is "Admin". Inherited from GroupChat. KeyBoardInterrupt will make the admin agent take over.
+    - func_call_filter: whether to enforce function call filter. Default is True. Inherited from GroupChat.
+        When set to True and when a message is a function call suggestion,
+        the next speaker will be chosen from an agent which contains the corresponding function name
+        in its `function_map`.
+    - allow_repeat_speaker: whether to allow the same speaker to speak consecutively. Default is True. Inherited from GroupChat.
+    """
+
+    def __init__(
+        self,
+        agents: List[Agent],
+        messages: List[Dict],
+        graph: nx.DiGraph,
+        max_round: int = 10,
+        admin_name: str = "Admin",
+        func_call_filter: bool = True,
+        allow_repeat_speaker: bool = True,
+    ):
+        # Inherit from GroupChat, and initialize with the given parameters (except graph)
+        super().__init__(
+            agents=agents,
+            messages=messages,
+            max_round=max_round,
+            admin_name=admin_name,
+            func_call_filter=func_call_filter,
+            speaker_selection_method="graph",
+            allow_repeat_speaker=allow_repeat_speaker,
+        )
+
+        self.previous_speaker = None  # Keep track of the previous speaker
+        self.graph = graph  # The graph depicting who are the next speakers available
+
+        # Check that the graph is a DiGraph
+        if not isinstance(self.graph, nx.DiGraph):
+            raise ValueError("The graph must be a networkx DiGraph.")
+
+        def _check_graph_validity(self):
```

Comment:
[Optional Comment]: 
This function is defined with the __init__ function. Why not move it outside the constructor?
For instance, maybe in the future the users can add more nodes into the graph, and then we will need "_check_graph_validity" again.

Reply:
Great idea with the mutable graph! I have now exposed `check_graph_validity` in the class.